src/NHibernate/Type/CollectionType.cs

```diff
@@ -391,14 +392,36 @@ public virtual object InstantiateResult(object original)
 			// in which case we should never ever ever reset the dirty flag
 			// on the target because we simply do not know...
 			IPersistentCollection originalPc = original as IPersistentCollection;
-			IPersistentCollection resultPc = result as IPersistentCollection;
-			if(originalPc != null && resultPc!=null)
+			IPersistentCollection resultPc = target as IPersistentCollection;
+			if(resultPc!=null)
 			{
-				if (!originalPc.IsDirty)
+				if (elementsEqual || (originalPc != null && !originalPc.IsDirty))
 					resultPc.ClearDirty();
 			}
 
-			return result;
+			return target;
+		}
+
+		protected virtual bool AreCollectionElementsEqual(IEnumerable original, IEnumerable target)
```

Comment:
>@hazzik, too much added complexity?

No. We either do it properly for all collection types or not do it. Why do we favor Sets over other collection types?